### ISSUE TITLE
tf.contrib.distributions.NegativeBinomial numerical stability (inf)

### ISSUE THREAD

**Issue by EdeMeijer** [tensorflow/tensorflow#12371]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.2.1
- **Python version**: 3
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem

I'm working on a negative binomial regression model and ran into `inf`/`nan` values during training. I was able to trace it back to the NegativeBinomial implementation, and specificially when dealing with relatively large probability logits.

Example to demonstrate the issue:
```python
import tensorflow as tf

logits = tf.constant([1.0, 5.0, 10.0, 15.0, 16.0, 17.0, 20.0, 50.0])
nb = tf.contrib.distributions.NegativeBinomial(10.0, logits)

sess = tf.Session()
x = sess.run(nb.log_prob(10.0))
print(x)
```
Results in
```
[  -4.83159161  -38.70070267  -88.56268311 -137.00408936 -147.99020386
          -inf          -inf          -inf]
```

Looking at the values before going to `-inf` (somewhere between 16.0 and 17.0) it doesn't seem to approaching any sort of limit, so I figured something else was going on.

Inside the `NegativeBinomial` class, the provides logits are first transformed to real probabilities by passing them through a sigmoid, which is `1.0 / (exp(-x) + 1.0)`. Later, in `_log_unnormalized_prob`, these probabilities go through a `log1p(-x)` function. The problem is that for big values and limited floating point precision, sigmoid will output exactly 1.0, which results in taking the log of -1.0 + 1.0, and the log of 0 yields negative infinity. 

I first tried to fix it by simplifying the combination of functions and getting rid of the sigmoid.

I replaced ` math_ops.log1p(-self.probs)` with `tf.log(1.0 / (tf.exp(self.logits) + 1.0))`, which is equivalent (`log((1 / (1 + exp(-x)) + 1) = log(1 / (exp(x) + 1))` but gives better numerical stability. For the same script but with higher values (1.0, 5.0, 10.0, 15.0, 16.0, 17.0, 20.0, 50.0, 70.0, 80.0, 90.0), the output now is:
```
[  -4.8315897   -38.70066452  -88.567276   -138.56636047 -148.56636047
 -158.56636047 -188.56636047 -488.56634521 -688.56634521 -788.56634521
          -inf]
```

This is better, but it still goes to -inf. Finally, I inspected the raw values of the function I changed, and noticed that it converges to `-x` somewhere after 15.0 as input, so my final solution looks like this:
```python
tf.where(self.logits <= 20.0, x=tf.log(1.0 / (tf.exp(self.logits) + 1.0)), y=-self.logits)
```

So ultimately I changed from
```python
def _log_unnormalized_prob(self, x):
    if self.validate_args:
        x = distribution_util.embed_check_nonnegative_integer_form(x)
    return (self.total_count * math_ops.log1p(-self.probs)
            + x * math_ops.log(self.probs))
```
to
```python
def _log_unnormalized_prob(self, x):
    if self.validate_args:
        x = distribution_util.embed_check_nonnegative_integer_form(x)
    return (self.total_count
            * tf.where(self.logits <= 20.0, x=tf.log(1.0 / (tf.exp(self.logits) + 1.0)), y=-self.logits)
            + x * math_ops.log(self.probs))
```

which gives the correct probabilities for arbitrarily large probability logits.

I don't feel like creating a PR since I'm not convinced this is the most elegant solution, and not sure whether this problem might exist at other placed in the code base, for example other distribution types. For now, I made my own copy of BinomialDistribution with this fix included, and I'd like to invite you to verify the issue, check the math and perhaps come up with a better solution.

Thanks.

**Comment by reedwm** [comment]
@jvdillon can you comment?

**Comment by EdeMeijer** [comment]
I ran into a similar but different issue with large negative logits. In this case, `probs` goes to 0.0 and the last log term of the expression goes to -inf. Similarly to the term I updated, the log converges to `logits`, so this case can be solved with a similar `where` op, like this.

```python
def _log_unnormalized_prob(self, x):
    if self.validate_args:
        x = distribution_util.embed_check_nonnegative_integer_form(x)
    return (self.total_count
            * tf.where(self.logits <= 20.0, x=tf.log(1.0 / (tf.exp(self.logits) + 1.0)), y=-self.logits)
            + x * tf.where(self.logits >= -20.0, x=tf.log(self.probs), y=self.logits))
```

**Comment by jvdillon** [comment]
Sorry for just now seeing this. Ill push this change right now.

**Comment by EdeMeijer** [comment]
That's okay, I'm fine with my own version until this lands.

**Comment by jvdillon** [comment]
Fix submitted using the following logic:

  def _log_unnormalized_prob(self, x):
    if self.validate_args:
      x = distribution_util.embed_check_nonnegative_integer_form(x)
    return (self.total_count * math_ops.log_sigmoid(-self.logits)
            + x * math_ops.log_sigmoid(self.logits))

**Comment by EdeMeijer** [comment]
Nice, I knew there had to be a much better way. `log(sigmoid(x))` is just `-softplus(-x)`, and I can see in `softplus_op.h` that it deals with the same underflow / overflow issue I tried to solve here, but of course much more efficiently. Thanks!
